src/eval.rs

```diff
@@ -0,0 +1,216 @@
+                    (Sub, Float(x), Integer(y)) => Float(x - *y as f64),
+
+                    (Mul, Integer(x), Integer(y)) => Integer(x * y),
+                    (Mul, Float(x), Float(y)) => Float(x * y),
+                    (Mul, Integer(x), Float(y)) => Float(*x as f64 * y),
+                    (Mul, Float(x), Integer(y)) => Float(x * *y as f64),
+
+                    (Div, Integer(x), Integer(y)) => Integer(
+                        x.checked_div(*y)
+                            .ok_or_else(|| EvalError::Unexpected("division by zero".to_string()))?,
+                    ),
+                    (Div, Float(x), Float(y)) => Float(x / y),
+                    (Div, Integer(x), Float(y)) => Float(*x as f64 / y),
+                    (Div, Float(x), Integer(y)) => Float(x / *y as f64),
+
+                    // It seems like the Nix reference implementation compares floats exactly
+                    // https://github.com/NixOS/nix/blob/4a5aa1dbf6/src/libutil/comparator.hh#L26
+                    (Equal, Integer(x), Integer(y)) => Bool(x == y),
+                    (Equal, Float(x), Float(y)) => Bool(x == y),
+                    (Equal, Integer(x), Float(y)) => Bool(*x as f64 == *y),
+                    (Equal, Float(x), Integer(y)) => Bool(*x == *y as f64),
+                    (Equal, Bool(x), Bool(y)) => Bool(x == y),
+
+                    (NotEqual, Integer(x), Integer(y)) => Bool(x != y),
+                    (NotEqual, Float(x), Float(y)) => Bool(x != y),
+                    (NotEqual, Integer(x), Float(y)) => Bool(*x as f64 != *y),
+                    (NotEqual, Float(x), Integer(y)) => Bool(*x != *y as f64),
+                    (NotEqual, Bool(x), Bool(y)) => Bool(x != y),
+
+                    (Less, Integer(x), Integer(y)) => Bool(x < y),
+                    (Less, Float(x), Float(y)) => Bool(x < y),
+                    (Less, Integer(x), Float(y)) => Bool((*x as f64) < *y),
+                    (Less, Float(x), Integer(y)) => Bool(*x < *y as f64),
+
+                    (LessOrEq, Integer(x), Integer(y)) => Bool(x <= y),
+                    (LessOrEq, Float(x), Float(y)) => Bool(x <= y),
+                    (LessOrEq, Integer(x), Float(y)) => Bool(*x as f64 <= *y),
+                    (LessOrEq, Float(x), Integer(y)) => Bool(*x <= *y as f64),
+
+                    (Greater, Integer(x), Integer(y)) => Bool(x > y),
+                    (Greater, Float(x), Float(y)) => Bool(x > y),
+                    (Greater, Integer(x), Float(y)) => Bool(*x as f64 > *y),
+                    (Greater, Float(x), Integer(y)) => Bool(*x > (*y as f64)),
+
+                    (GreaterOrEq, Integer(x), Integer(y)) => Bool(x >= y),
+                    (GreaterOrEq, Float(x), Float(y)) => Bool(x >= y),
+                    (GreaterOrEq, Integer(x), Float(y)) => Bool(*x as f64 >= *y),
+                    (GreaterOrEq, Float(x), Integer(y)) => Bool(*x >= (*y as f64)),
```

Comment:
I'm wondering if it'd make sense to put this four expressions for a binary operator into some kind of macro :)

Reply:
Good idea!